allennlp/tests/data/transforms/basic_iterator_transforms_compat_test.py

```diff
@@ -0,0 +1,295 @@
+from typing import List, Iterable, Dict, Union
```

Comment:
This is pretty much just a copy of the existing test, right? As a reader, it's a bit hard to see that as-is. And we'll also be liable to drift. Would it be possible to reduce this duplication by changing the existing tests to call a factory function that either uses the old class or the new stub? Then we could parameterize the test to use the factory function in one way or the other.

Reply:
I'm going to leave this as is, because 1) It's not too much duplication, 2) immediately after merging this PR i'll open a PR to depreciate the old iterators, and it will be easier to completely remove.

There is additionally a functional difference, which is that a lot of the iterator tests use the private method `_create_batches` to generate the batches in the tests - the new iterator actually yields `TensorDicts` here, because the operation is expensive and should be pushed inside the dataloader workers. So to test, I monkeypatched that function to _not_ call `batch.as_tensor_dict()`, so I could test the new iterator fulfilled the spec of the old iterator. Some of the tests require that patch and some don't, which made parameterising the tests a bit messy. Because of this I think it's easier to have them separate.

Reply:
My concern isn't one of code purity or anything. It's a practical question, "How do we avoid regressions when using different tests?" Even though my prior is that you did a great job on these tests, it's really hard to actually help as a reviewer with a big diff like this. (And personally, I'd have difficulty ensuring I didn't make some mistake when converting the tests in this manner.) Another issue is that given that we want to support people who may have subclassed the old iterators, the deprecation may not move particularly quickly. I'd argue that this sort of thing is really worth spending some time on.

Thanks for pointing out the monkey patch issue, that's definitely complicates things. Could you add some comments around why that's necessary? I'll take a stab at parameterizing things. Have a rough idea of something that might work...

Reply:
Thanks - I don't need you to take a stab at parameterizing it, I actually had it similar to what you are suggesting before: https://github.com/allenai/allennlp/pull/3386/commits/770999c98b34dfec51cb8f0ea360373cb630e9b5

I did it like this because I actually think it's better. Regardless, we need to iron out whether we are removing these old iterators before I change this.

Feel a bit stuck in between a rock and a hard place here - if i'd have changed the tests for the current iterators, your comment would have been that it's unclear if backward compatibility is maintained.

Reply:
Sorry, my intent is definitely not to make you feel stuck. Hence why I should try to parameterize to demonstrate what I'm talking about. My goal is to do something more surgical such that it's clear that backwards compatibility is maintained. (By the way, was https://github.com/allenai/allennlp/commit/770999c98b34dfec51cb8f0ea360373cb630e9b5 really the one you meant? Doesn't seem parameterized.)